focus/operations/src/pull.rs

```diff
@@ -0,0 +1,210 @@
+        bail!("Could not find prefetch ref or ref is empty");
+    }
+
+    Ok(())
+}
+
+/// Validate that `prefetch` refs are ahead of current HEAD, and
+/// validate that `focus/sync` ref is ahead of current HEAD
+fn validate_merge_base(app: Arc<App>, repo_path: &Path) -> Result<()> {
+    let default_prefetch_ref_sha =
+        git_helper::parse_ref(app.clone(), repo_path, PREFETCH_DEFAULT_REF)
+            .expect("Could not parse default prefetch ref");
+    let focus_sync_ref_sha = git_helper::parse_ref(app.clone(), repo_path, FOCUS_SYNC_REF)
+        .expect("Could not parse `refs/focus/sync`");
+    let current_head = git_helper::get_current_revision(app.clone(), repo_path)?;
+
+    // If prefetch refs and HEAD are equal then we can exit early
+    if default_prefetch_ref_sha == current_head {
+        bail!("HEAD is up to date with prefetch, nothing to do.")
+    }
+
+    let merge_base_prefetch_and_head = git_helper::get_merge_base(
+        app.clone(),
+        repo_path,
+        &current_head,
+        PREFETCH_DEFAULT_REF,
+        None,
+    )
+    .context(
+        "Could not get merge-base between current HEAD and 'refs/prefetch/remotes/origin/master'",
+    )?;
+
+    // If prefetch is behind current HEAD, then exit early
+    if merge_base_prefetch_and_head == default_prefetch_ref_sha {
+        bail!("Exiting: Prefetch is behind HEAD");
+    }
+
+    // Tests fail if I take out the redundant clone here
+    #[allow(clippy::redundant_clone)]
+    let merge_base_focus_sync_and_head =
+        git_helper::get_merge_base(app.clone(), repo_path, &current_head, FOCUS_SYNC_REF, None)
+            .context("Could not get merge-base between current HEAD and 'refs/focus/sync'")?;
+
+    // If focus sync is behind current HEAD, then exit early
+    if merge_base_focus_sync_and_head == focus_sync_ref_sha {
+        bail!("Exiting: refs/focus/sync is behind HEAD");
+    }
+
+    Ok(())
+}
+
+fn preflight_checks(app: Arc<App>, repo_path: &Path) -> Result<()> {
+    validation_ref_existence(app.clone(), repo_path)?;
+    validate_merge_base(app, repo_path)?;
+
+    Ok(())
+}
+
+fn update_refs_from_prefetch(app: Arc<App>, repo_path: &Path) -> Result<()> {
+    // TODO: don't assume remote names
+    git_helper::fetch_refs(
+        repo_path,
+        ["+refs/prefetch/remotes/origin/*:refs/remotes/origin/*"].iter(),
+        ".",
+        app,
+        None,
+    )
+    .context("Fetching from prefetch")?;
+
+    Ok(())
+}
+
+/// Update the current branch from `refs/focus/sync`
+fn update_current_branch(app: Arc<App>, repo_path: &Path) -> Result<()> {
+    let current_branch = git_helper::get_current_branch(app.clone(), repo_path)
+        .context("Could not get current branch")?;
+    if current_branch.is_empty() {
+        bail!("HEAD does not point to a branch");
+    }
+    git_helper::pull(
+        repo_path,
+        [format!("+refs/focus/sync:refs/{}", current_branch)].iter(),
```

Comment:
Not sure if the + makes sense here...
https://git-scm.com/book/en/v2/Git-Internals-The-Refspec

Reply:
Yep, you're right. fixing